matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/send/LocalEchoEventFactory.kt

```diff
@@ -58,37 +60,68 @@ internal class LocalEchoEventFactory @Inject constructor(@UserId private val use
     // TODO Inject
     private val renderer = HtmlRenderer.builder().build()
 
-    fun createTextEvent(roomId: String, msgType: String, text: String, autoMarkdown: Boolean): Event {
-        if (msgType == MessageType.MSGTYPE_TEXT) {
-            return createFormattedTextEvent(roomId, createTextContent(text, autoMarkdown))
+    fun createTextEvent(roomId: String, msgType: String, text: CharSequence, autoMarkdown: Boolean): Event {
+        if (msgType == MessageType.MSGTYPE_TEXT || msgType == MessageType.MSGTYPE_EMOTE) {
+            return createFormattedTextEvent(roomId, createTextContent(text, autoMarkdown), msgType)
         }
-        val content = MessageTextContent(type = msgType, body = text)
+        val content = MessageTextContent(type = msgType, body = text.toString())
         return createEvent(roomId, content)
     }
 
-    private fun createTextContent(text: String, autoMarkdown: Boolean): TextContent {
+    private fun createTextContent(text: CharSequence, autoMarkdown: Boolean): TextContent {
         if (autoMarkdown) {
-            val document = parser.parse(text)
+            val source = transformPills(text, "[%2\$s](https://matrix.to/#/%1\$s)")
+                    ?: text.toString()
+            val document = parser.parse(source)
             val htmlText = renderer.render(document)
 
-            if (isFormattedTextPertinent(text, htmlText)) {
-                return TextContent(text, htmlText)
+            if (isFormattedTextPertinent(source, htmlText)) {
+                return TextContent(source, htmlText)
+            }
+        } else {
+            // Try to detect pills
+            transformPills(text, "<a href=\"https://matrix.to/#/%1\$s\">%2\$s</a>")?.let {
+                return TextContent(text.toString(), it)
             }
         }
 
-        return TextContent(text)
+        return TextContent(text.toString())
+    }
+
+    private fun transformPills(text: CharSequence,
```

Comment:
This method should be in a dedicated class, along with the 2 templates

Reply:
Moved to TextPillsUtils.kt